### ISSUE TITLE
phpcs: Unexpected end of JSON input

### ISSUE THREAD

**Issue by ilidan85** [ikappas/vscode-phpcs#42]
Windows 10 Home
VS Code v. 1.11.2

Plugin not working on opened file. After saving showing notification with error: 
"phpcs: Unexpected end of JSON input"

It worked perfectly yesterday.

**Comment by mishac** [comment]
I'm having the same issue.  Pulling my hair out trying to figure out why it's happening!

**Comment by christophherr** [comment]
Just in case, here is the callstack I´m seeing:
```
phpcs: Unexpected end of JSON input           messageService.ts:126 
e.doShow @ messageService.ts:126
e.show @ messageService.ts:105
(anonymous function) @ mainThreadMessageService.ts:64
n.Class.derive._oncancel @ winjs.base.raw.js:1641
t.showMessage @ mainThreadMessageService.ts:34
t.$showMessage @ mainThreadMessageService.ts:28
e.handle @ abstractThreadService.ts:34
s @ ipcRemoteCom.ts:275
f @ ipcRemoteCom.ts:231
_combinedTickCallback @ internal/process/next_tick.js:67
_tickCallback @ internal/process/next_tick.js:98```

**Comment by christophherr** [comment]
If I open a single file directly (from the Windows explorer or from inside VSC with file - open file), I am getting the `phpcs: Unexpected end of JSON input` error.
http://prntscr.com/f2vc6z

If I open the folder containing the file and then click on the file / open the file, everything works.
http://prntscr.com/f2vckq

**Comment by drjonnicholson** [comment]
Same problem here. If I open code within the workspace folder it works perfectly. If I open a file from that workspace in a new window (e.g. drag an editor from one instance to another on a different monitor) then I get this error.

If I open the folder above my workspace folder as a new workspace for the second vscode instance then the error goes away, but  of course it's now using the default code standard rather than the one for our projects.

**Comment by seanvree** [comment]
Same issue here. So annoying.

**Comment by jchristopher** [comment]
I hit this issue just now but after receiving some help I was clued into the underlying issue being my version of `phpcs`. I had `2.x` installed because it was a requirement of [the standard I was using](https://github.com/10up/10up-code-review) but after removing that and updating `phpcs` to `3.x` the error message is gone and I'm able to use another standard for the time being.

**Comment by kendinitto** [comment]
I am having the exact same problem on my laptop. If I open a file directly in Visual Studio Code I get "phpcs: Unexpected end of JSON input". If I open a project folder first then open the file from within that project folder it works fine. This is a tough nut to solve. I want to be able to open a file directly because that way I can use flashfxp to edit files via sftp with VScode. VScode does not have native sftp edit support and all the extensions I have tried do not work well. Notepad++ has great sftp file edit functionality. On my desktop I do not have this problem, only on my laptop. On my desktop i can edit files directly with VScode and phpcs no problem. Both systems are configured the exact same way Windws 10, same version of php, phpcs, (3.1) and visual code. Same ENV variables etc... If anyone can help would be greatly appreciated. Thanks in advance.

**Comment by ikappas** [comment]
Hi everyone and thank you for reporting this issue. 

Can you specify which version of the plugin you are using? 

The development branch has significant changes that may help. Can you please test whether the issues you are experiencing are resolved with the latest version?

**Comment by arijnr** [comment]
I have the same problem:
![issue](https://user-images.githubusercontent.com/14283324/32191736-3392e4c4-bd99-11e7-8996-435c5a57e12c.png)

**Comment by ikappas** [comment]
I just pushed some major new features in `develop` branch that need testing and include: 

- ability to set phpcs.executablePath
- ability to set phpcs.showWarnings 
- ability to set phpcs.showSources 
- ability to enable phpcs.trace.server with messages or verbose option for tracing errors.
- ability to set phpcs.composerJsonPath to something other that workspaceRoot/composer.json

Your feedback is very welcome!

**Comment by ikappas** [comment]
This issue should be resolved in latest `develop` branch so please consider testing your use cases.

**Comment by arijnr** [comment]
I tested the latest develop branch and the issue is solved in my machine, thank you!

**Comment by christophherr** [comment]
Just tested the latest develop branch and it worked with single php files. 
No "Unexpected end of JSON input" errors.

I had to set the phpcs.executablePath (error message was something along the lines of "expected String, received null").

**Comment by ikappas** [comment]
@arijnr @christophherr Thank you for taking some time to test the new version!

@christophherr can you temporarily disable the `"phpcs.executablePath"` and set `"phpcs.trace.server" : "messages"` config in you settings and post back your trace from `PHP Code Sniffer` output to further debug the expected String, received null error?

**Comment by christophherr** [comment]
@ikappas 
Unfortunately, there is not Output for PHP Code Sniffer. 
If I set `"phpcs.trace.server": "verbose"` 
the output is:
```
[Trace - 11:46:27] Received request 'window/showMessageRequest - (3)'.
Params: {
    "type": 1,
    "message": "phpcs: Path must be a string. Received null",
    "actions": []
}
```
Did the previous version scan/search for the executable?
(I installed phpcs globally with Composer)

**Comment by ikappas** [comment]
@christophherr it still does scan/search for the executable. `phpcs.executablePath` bypasses this. Do you have C:\Users\{ your username here }}\AppData\Roaming\Composer\vendor\bin in your path (basically the path reported by the command: `composer global config bin-dir --absolute`)

**Comment by christophherr** [comment]
Yes, it's in the path.
Not sure if it matters but there is a "slash mismatch" at the vendor and vendor\bin folders 
using `composer global config --list` but the slashes "match" with `where phpcs` and `which phpcs`

![slash-mismatch](https://user-images.githubusercontent.com/8683126/32409750-e91d3d84-c187-11e7-9aee-e4bf2a2afa8b.jpg)
 
It works setting `"phpcs.composerJsonPath"`  or `"phpcs.executablePath"`.

**Comment by ikappas** [comment]
@christophherr I can't reproduce your issue in a new windows installation from scratch.

**Comment by kendinitto** [comment]
Hello, I want to try the develop branch but I usually just use pear to install it , "pear install PHP_CodeSniffer" is there a way to install it with pear or is git the only way?

**Comment by phproberto** [comment]
Same issue here. I tried to follow instructions in the [README](https://github.com/ikappas/vscode-phpcs/blob/develop/README.md#setup-development-version) to load development version but `npm install` task does not exist. I've tried to guess the setup to contribute an update for the README but I was not able to.

System info:
* **OS:** Ubuntu 16.04
* **phpcs version:** `PHP_CodeSniffer version 2.9.1 (stable)` installed as composer dependency.
* **local phpcs command that works:** `./vendor/bin/phpcs`

For now I've uninstalled it :(

**Comment by cbejensen** [comment]
I had the `Unexpected end of JSON input` issue as well, but following the instructions for setting up the development version resolved it.

**Comment by zepdesign** [comment]
Forgive my ignorance but I installed PHP_CodeSniffer via pear. Is there a way to swap to a development version there or when is there likely to be a fixed release?

**Comment by Morgy93** [comment]
@ikappas Is this not fixed in `master` yet?
Still getting the error with v0.7.0 from VSCode extension marketplace.

**Comment by ikappas** [comment]
I've just pushed 1.0.0-beta.6 on the development branch adding automatic configuration search. This is the last feature for the 1.0.0 release which needs some testing before 1.0.0 launch.

**Comment by ikappas** [comment]
I just released vscode-phpcs 1.0.0 which supports the VSCode 1.20+ and includes many enhancements most notable of which are:

1. multi root support
2. lint as you type support
3. automatic configuration search support
4. ability to set phpcs.executablePath
5. ability to set phpcs.showWarnings
7. ability to set phpcs.showSources
8. ability to set phpcs.composerJsonPath to something other that workspaceRoot/composer.json
9. ability to enable phpcs.trace.server with messages or verbose option for tracing errors.

I would like to give my thanks to all the people that helped track down issues in the previous betas and hope the new features help you get things done easier.

On that note, I am closing this issue to properly track any new issues in the new release. If you find this issue persists or any other issue please feel free to open a new ticket for v.1.0.0.
